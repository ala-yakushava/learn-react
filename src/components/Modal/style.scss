@import "../../styles/theme";

.Modal {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(black, 0.6);
  overflow-y: auto;

  &_wrapper {
    position: absolute;
    left: 50%;
    width: 680px;
    margin: 100px 0;
    padding: 70px 70px 70px 90px;
    color: $white;
    background-color: $darkest-grey;
    transform: translate(-50%);
  }

  &_close {
    @include reset-button;
    position: absolute;
    top: 10px;
    right: 10px;
    padding: 15px;
    color: $white;
  }
};
