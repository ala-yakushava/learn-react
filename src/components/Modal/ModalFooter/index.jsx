import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';

const ModalFooter = ({ children }) => (
  <div className="ModalFooter">
    { children }
  </div>
);

ModalFooter.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ModalFooter;
