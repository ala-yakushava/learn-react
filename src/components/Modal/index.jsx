import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import './style.scss';

const Modal = ({ children, onClick }) => {
  const modalRoot = document.getElementById('modal-root');
  const element = document.createElement('div');

  useEffect(() => {
    modalRoot.appendChild(element);
    document.body.style.overflow = 'hidden';
    document.body.style.paddingRight = '17px';
    return () => {
      modalRoot.removeChild(element);
      document.body.style.overflow = 'auto';
      document.body.style.paddingRight = 0;
    };
  });

  return ReactDOM.createPortal(
    <div className="Modal">
      <div className="Modal_wrapper">
        <button type="button" className="Modal_close" onClick={onClick}>X</button>
        {children}
      </div>
    </div>,
    element,
  );
};

Modal.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func,
};

Modal.defaultProps = {
  onClick: () => {},
};

export default Modal;
