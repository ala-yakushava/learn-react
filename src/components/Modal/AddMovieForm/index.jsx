import React from 'react';

import './style.scss';
import Heading from '../../Heading';
import TextInput from '../../TextInput';
import ModalFooter from '../ModalFooter';
import Button from '../../Button';
import { formField } from '../../../data';

class AddMovieForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      release_date: '',
      url: '',
      genres: '',
      overview: '',
      runtime: '',
    };
  }

  handleInputChange = () => (evt) => {
    const { target: { name, value } } = evt;

    this.setState({
      [name]: value,
    });
  }

  handleSubmit = (evt) => {
    console.log(this.state);
    evt.preventDefault();
  }

  handleReset = () => {
    this.setState({
      title: '',
      release_date: '',
      url: '',
      genres: '',
      overview: '',
      runtime: '',
    })
  }

  render() {
    return (
      <form
        className="AddMovieForm"
        onSubmit={this.handleSubmit}
        onReset={this.handleReset}
      >
        <Heading>Add Movie</Heading>
        {formField.map(({ name, label, placeholder}, i) => (
          <TextInput
            key={i}
            name={name}
            label={label}
            placeholder={placeholder}
            value={this.state[name]}
            isRequired={true}
            onChange={this.handleInputChange(this.state[name])}
          />
        ))}
        <ModalFooter>
          <Button type="reset" mode="secondary">Reset</Button>
          <Button type="submit" mode="primary">Submit</Button>
        </ModalFooter>
      </form>
    );
  }
}

export default AddMovieForm;
