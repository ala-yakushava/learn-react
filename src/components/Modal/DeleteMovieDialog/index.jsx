import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';
import Heading from '../../Heading';
import ModalFooter from '../ModalFooter';
import Button from '../../Button';

const DeleteMovieDialog = ({ onClick }) => (
  <div className="DeleteMovieDialog" data-testid="delete-dialog">
    <Heading>Delete Movie</Heading>
    <p>Are you sure you want to delete this movie?</p>
    <ModalFooter>
      <Button mode="primary" onClick={onClick}>Confirm</Button>
    </ModalFooter>
  </div>
);

DeleteMovieDialog.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default DeleteMovieDialog;
