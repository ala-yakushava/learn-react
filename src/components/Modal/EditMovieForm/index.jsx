import React, { useState } from 'react';
import PropTypes from 'prop-types';

import './style.scss';
import Heading from '../../Heading';
import TextInput from '../../TextInput';
import ModalFooter from '../ModalFooter';
import Button from '../../Button';
import { formField } from '../../../data';

const EditMovieForm = ({ movie, onSubmit }) => {
  const defaultState = {
    title: movie.title,
    release_date: movie.release_date,
    poster_path: movie.poster_path,
    genres: movie.genres.join(', '),
    overview: movie.overview,
    runtime: String(movie.runtime),
  };

  const [form, setValue] = useState(defaultState);

  const handleReset = () => setValue(defaultState);

  const handleSubmit = (evt) => {
    const data = {
      ...form,
      id: movie.id,
      runtime: Number(form.runtime),
      genres: form.genres.split(', '),
    };

    evt.preventDefault();
    onSubmit(data);
  };

  const handleInputChange = () => (evt) => {
    const { name, value } = evt.target;
    setValue({ ...form, [name]: value });
  };

  return (
    <form
      className="EditMovieForm"
      onSubmit={handleSubmit}
      onReset={handleReset}
    >
      <Heading>Edit Movie</Heading>
      <TextInput
        value={String(movie.id)}
        readOnly
        label="Movie ID"
      />
      {formField.map(({
        id, name, label, placeholder,
      }) => (
        <TextInput
          key={id}
          name={name}
          label={label}
          placeholder={placeholder}
          value={form[name]}
          required
          onChange={handleInputChange(form[name])}
        />
      ))}
      <ModalFooter>
        <Button type="reset" mode="secondary">Reset</Button>
        <Button type="submit" mode="primary">Save</Button>
      </ModalFooter>
    </form>
  );
};

EditMovieForm.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    genres: PropTypes.arrayOf(PropTypes.string),
    release_date: PropTypes.string,
    poster_path: PropTypes.string.isRequired,
    overview: PropTypes.string.isRequired,
    runtime: PropTypes.number.isRequired,
  }).isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export default EditMovieForm;
