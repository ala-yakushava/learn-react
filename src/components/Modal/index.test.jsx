import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import Modal from './index';

describe('Components - Modal', () => {
  const Child = () => <div>Forms</div>;
  const handleClick = jest.fn();

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => element);

    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    const body = document.querySelector('body');
    body.appendChild(modalRoot);
  });

  afterAll(() => {
    ReactDOM.createPortal.mockClear();
  });

  test('should match snapshot', () => {
    const tree = renderer
      .create(
        <Modal onClick={handleClick}>
          <Child />
        </Modal>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should mount modal to document', () => {
    const { getByTestId, getByRole } = render(
      <Modal onClick={handleClick}>
        <Child />
      </Modal>,
    );

    expect(getByTestId('modal')).toBeInTheDocument();

    userEvent.click(getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
