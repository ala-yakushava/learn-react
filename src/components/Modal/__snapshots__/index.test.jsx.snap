// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Modal should match snapshot 1`] = `
<div
  className="Modal"
  data-testid="modal"
>
  <div
    className="Modal_wrapper"
  >
    <button
      className="Modal_close"
      onClick={[MockFunction]}
      type="button"
    >
      X
    </button>
    <div>
      Forms
    </div>
  </div>
</div>
`;
